src/transactions/TransactionUtils.cpp

```diff
@@ -1172,6 +1174,321 @@ removeOffersByAccountAndAsset(AbstractLedgerTxn& ltx, AccountID const& account,
+        // TrustLineEntry
+        auto poolTL = [&]() -> TrustLineEntry const& {
+            return poolShareTrustLine.current().data.trustLine();
+        };
+
+        auto poolID = poolTL().asset.liquidityPoolID();
+        auto pool = loadLiquidityPool(ltxInner, poolID);
+        auto balance = poolTL().balance;
+        auto cbSponsoringAccID = getTrustLineBacker(poolShareTrustLine);
+
+        // release reserves for pool share trustline that will be deleted
+        // later
+        {
+            auto trustAcc = stellar::loadAccount(ltxInner, accountID);
+            removeEntryWithPossibleSponsorship(ltxInner, ltxInner.loadHeader(),
+                                               poolShareTrustLine.current(),
+                                               trustAcc);
+
+            // using poolTL() after this will throw an exception
+            poolShareTrustLine.erase();
+        }
+
+        auto redeemIntoClaimableBalance =
+            [&ltxInner, &txSourceID, txSeqNum, opIndex, &poolID, &accountID,
+             &cbSponsoringAccID](Asset const& assetInPool,
+                                 int64_t amount) -> bool {
+            // if the claimant is the issuer, then we don't create the claimable
+            // balance
+            if (isIssuer(accountID, assetInPool))
+            {
+                return true;
+            }
+
+            // create a claimable balance for the asset in the pool
+            LedgerEntry claimableBalance;
+            claimableBalance.data.type(CLAIMABLE_BALANCE);
+
+            auto& claimableBalanceEntry =
+                claimableBalance.data.claimableBalance();
+
+            claimableBalanceEntry.balanceID =
+                getRevokeID(txSourceID, txSeqNum, opIndex, poolID, assetInPool);
+            claimableBalanceEntry.amount = amount;
+            claimableBalanceEntry.asset = assetInPool;
+            claimableBalanceEntry.claimants = {
+                makeUnconditionalClaimant(accountID)};
+
+            // if this asset isn't native
+            // 1. set clawback if it's set on the trustline
+            // 2. decrement liquidityPoolUseCount on the asset trustline
+            if (assetInPool.type() != ASSET_TYPE_NATIVE)
+            {
+                // trustline must exist
+                auto assetTrustLine =
+                    loadTrustLine(ltxInner, accountID, assetInPool);
+                if (assetTrustLine.isClawbackEnabled())
+                {
+                    setClaimableBalanceClawbackEnabled(claimableBalanceEntry);
+                }
+            }
+
+            // The account that is currently sponsoring the soon to be
+            // deleted pool share trustline will be used for the claimable
+            // balances. If that account is currently in a sponsorship
+            // sandwich, that sponsoring account will instead sponsor the
+            // claimable balance.
+            auto cbSponsoringLtxAcc =
```

Comment:
this is a bit sad that it's basically copy/pasted from `CreateClaimableBalanceOpFrame`

Reply:
I'm not sure if it's worth unifying the two blocks. They are different enough where I think the code would just get more complex.